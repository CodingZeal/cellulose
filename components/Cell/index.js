import React, { PureComponent } from 'react'
import { themr } from 'react-css-themr'
import {
  defaultTo,
  F,
  ifElse,
  is,
  lensProp,
  over,
  pick,
  prop,
  when
} from 'ramda'
import classnames from 'classnames'

import cellTheme from './theme.scss'

export class Cell extends PureComponent {
  flexProperties = [
    'alignSelf',
    'flexBasis',
    'flexGrow',
    'flexShrink',
    'flex',
    'order'
  ]

  render() {
    const { theme, breakPoint, children, columns } = this.props
    const spanOptions = over(
      lensProp('0'),
      defaultTo(columns),
      this.props.spanOptions
    )
    const span = prop(breakPoint, spanOptions)
    const cols = when(is(Object), prop('cols'))(span)
    const className = ifElse(is(Object), prop('className'), F)(span)
    const flexStyles = pick(this.flexProperties)(this.props)
    const width = `${cols / columns * 100}%`
    const cellClassName = classnames(
      theme.cell,
      { [className]: className }
    )

    return (
      <div style={{ ...flexStyles, width }} className={cellClassName}>
        {children}
      </div>
    )
  }
}

export default themr('', cellTheme)(Cell)
