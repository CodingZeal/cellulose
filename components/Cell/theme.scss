.cell {
  box-sizing: border-box;
  flex: 0 0 auto;
}
