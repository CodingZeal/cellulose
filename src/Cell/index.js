import React, { PureComponent } from 'react'
import {
  defaultTo,
  F,
  ifElse,
  is,
  lensProp,
  over,
  pick,
  prop,
  when
} from 'ramda'
import classnames from 'classnames'
import styled from 'styled-components'

const StyledCell = styled.div`
  box-sizing: border-box;
  flex: 0 0 auto;
`
export default class Cell extends PureComponent {
  flexProperties = [
    'alignSelf',
    'flexBasis',
    'flexGrow',
    'flexShrink',
    'flex',
    'order'
  ]

  render() {
    const { breakPoint, children, columns } = this.props
    const spanOptions = over(
      lensProp('0'),
      defaultTo(columns),
      this.props.spanOptions
    )
    const span = prop(breakPoint, spanOptions)
    const cols = when(is(Object), prop('cols'))(span)
    const className = ifElse(is(Object), prop('className'), F)(span)
    const flexStyles = pick(this.flexProperties)(this.props)
    const width = `${cols / columns * 100}%`

    const cellClassName = classnames(
      { [className]: className }
    )

    return (
      <StyledCell style={{ ...flexStyles, width }} className={cellClassName}>
        {children}
      </StyledCell>
    )
  }
}
