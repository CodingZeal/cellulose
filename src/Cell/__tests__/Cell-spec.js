import React from 'react'
import { shallow } from 'enzyme'
import { has } from 'ramda'
import renderer from 'react-test-renderer'

import Cell from '../'

describe('Cell', () => {
  const wrapper = shallow(
    <Cell
        textAlign='center'
        alignSelf='center'
        spanOptions={{ 0: 12, 768: { cols: 6, className: 'col6' } }}
        columns={12}
        breakPoint={768}>
      <div>One</div>
    </Cell>
  )
  const style = wrapper.prop('style')

  test('it passes white listed flex props to style', () => {
    expect(has('alignSelf')(style)).toBe(true)
    expect(has('textAlign')(style)).toBe(false)
  })

  test('it computes the width based on colums and span', () => {
    expect(style.width).toBe('50%')
  })

  test('it adds col6 className', () => {
    expect(wrapper.hasClass('col6')).toBe(true)
  })

  test('cell renders correctly', () => {
    const tree = renderer.create(
      <Cell
          textAlign='center'
          alignSelf='center'
          spanOptions={{ 0: 12, 768: { cols: 6, className: 'col6' } }}
          columns={12}
          breakPoint={768}>
        <div>One</div>
      </Cell>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
