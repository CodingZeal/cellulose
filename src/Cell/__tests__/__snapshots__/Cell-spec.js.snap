exports[`Cell cell renders correctly 1`] = `
<div
  className="col6 kYySsF"
  style={
    Object {
      "alignSelf": "center",
      "width": "50%",
    }
  }>
  <div>
    One
  </div>
</div>
`;
